package com.atguigu.util;

import java.lang.reflect.Method;
import java.util.Arrays;

import org.aspectj.lang.annotation.After;
import org.aspectj.lang.annotation.AfterReturning;
import org.aspectj.lang.annotation.AfterThrowing;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.annotation.Before;
import org.springframework.stereotype.Component;

@Aspect
@Component
public class LogUtils {
	
	/**
	 * 切入点表达式：
	 * 固定格式: execution(访问权限夫 返回值类型 方法全类名（参数表）)
	 * 
	 * 通配符： 
	 * 	* 
	 * 		1.匹配一个或者多个字符：
	 * 			execution(public int com.atguigu.impl.MyMathCalculator.*(int, int))
	 * 		2. 匹配任意一个参数，
	 * 			execution(public int com.atguigu.impl.MyMathCalculator.*(int, *))
	 * 		3. 只能匹配一层路径
	 * 
	 * 
	 *  ..  1。 匹配任意多个参数或者任意多个字符；
	 *  		execution(public int com.atguigu.impl.MyMathCalculator.*(..))
	 * 		2. 匹配任意多层路径
	 */
	@Before("execution(public int com.atguigu.impl.MyMathCalculator.*(..))")
	public static void getMethodBefore(){
		System.out.println("");
	}
	
	@AfterReturning("execution(public int com.atguigu.impl.MyMathCalculator.*(int, int))")
	public static void getMethodAfter(Method method, Object... args){
		System.out.println(method.getName() +" 方法正在使用调用结束，参数为" + Arrays.asList(args));
	}
	
	@AfterThrowing("execution(public int com.atguigu.impl.MyMathCalculator.*(int, int))")
	public static void getMethodError(Method method, Exception e){
		System.out.println(method.getName() +" 方法正在使用异常，参数为" + e.getCause());
	}
	
	@After("execution(public int com.atguigu.impl.MyMathCalculator.*(int, int))")
	public static void getMethodFinally(Method method, Object... args){
		System.out.println(method.getName() +" 方法正在使用最终，参数为" + Arrays.asList(args));
	}
	

}
